Tue Apr 17 21:41:11 2001  Jeffrey A Law  (law@cygnus.com)

	* jump.c (mark_all_labels): Canonicalize the tail recursion
	label attached to CALL_PLACEHOLDER insns.

2001-08-24  Jakub Jelinek  <jakub@redhat.com>

	* gcc.c-torture/compile/20010824-1.c: New test.

--- gcc/testsuite/gcc.c-torture/compile/20010824-1.c.jj	Fri Aug 24 13:06:35 2001
+++ gcc/testsuite/gcc.c-torture/compile/20010824-1.c	Mon Aug 20 10:36:13 2001
@@ -0,0 +1,7 @@
+void f(int n)
+{
+bugcauser:
+  if (n != 0)
+    f(n-1);
+  return;
+}
--- gcc/jump.c.jj	Thu Aug  9 18:03:19 2001
+++ gcc/jump.c	Mon Aug 20 15:03:59 2001
@@ -863,6 +863,17 @@ mark_all_labels (f, cross_jump)
 	    mark_all_labels (XEXP (PATTERN (insn), 0), cross_jump);
 	    mark_all_labels (XEXP (PATTERN (insn), 1), cross_jump);
 	    mark_all_labels (XEXP (PATTERN (insn), 2), cross_jump);
+
+	    /* Canonicalize the tail recursion label attached to the
+	       CALL_PLACEHOLDER insn.  */
+	    if (XEXP (PATTERN (insn), 3))
+	      {
+		rtx label_ref = gen_rtx_LABEL_REF (VOIDmode,
+						   XEXP (PATTERN (insn), 3));
+		mark_jump_label (label_ref, insn, cross_jump, 0);
+		XEXP (PATTERN (insn), 3) = XEXP (label_ref, 0);
+	      }
+
 	    continue;
 	  }
 	
